### ISSUE TITLE
Client credentials grant type, client claims

### ISSUE THREAD

**Issue by a23o** [IdentityServer/IdentityServer3#1800]
I have dynamically registered clients in my system. Client secret and client id are generated for the user in a "linking" process. A subscription id claim is registered for the client as a client claim. I can see the client claim in the access token fine, the problem is that I can't access it in my resource server. I use IdentityServer.AccesTokenValidation to validate the scope, but I can't find any method to obtain the client claims. If I understand well, it is because the access token does not have a subject, so the System.Security.Principal.WindowsIdentity is empty. No user, no claims. 

How can I access the client claims in a katana middleware?

A bit more on what I want to achieve: a user installs a long running client application. Logs in with his credentials once. Registers the client to the server, but the client registration should be tied to a "subscription", not to the user. There should be a way to withdraw/block the client on the server side. Clients should be able to use resources with the subscription claim. The client credentials grant is the only one, which doesn't involve a user, that's why I opted to use dynamically generated client ids with client claims. 

Am I doing it wrong? 

Security is hard. :)

**Comment by brockallen** [comment]
We do have support for client claims (claims on the client configuration) -- have you looked into that?

**Comment by a23o** [comment]
Hi Brock, 

yes, I have set up IdentityServer to issue client claims, that's what I meant by " I can see the client claim in the access token fine". The client claims are in the access token. My problem is, that I can't get them on the resource server. I don't know which owin/katana middleware to use.

**Comment by leastprivilege** [comment]
use this one: 

https://github.com/IdentityServer/IdentityServer3.AccessTokenValidation

**Comment by a23o** [comment]
Indeed I am using that. Scope validation is working fine. My application framework is Nancy, but that should not be a concern, because I am just doing this:  IAuthenticationManager auth = ctx.GetAuthenticationManager();  and authenticationManager.User is a ClaimsPrincipal. But this ClaimsPrincipal doesn't have any claims.

**Comment by a23o** [comment]
I'll try to rephrase my question, because either I am really close to my solution, or I will have to redesign my whole system. So, can you tell me how to get the client claims with any owin middleware that can validate the token issued by Idsrv3?
Thanx.

**Comment by leastprivilege** [comment]
It's been a while since I used Nancy - but you can check if the claims are inside the token (e.g. using http://jwt.io) - if yes, it is a MW problem. This definitely works in Web API - and should in Nancy as well.

Maybe ask here as well: https://github.com/NancyFx/Nancy.MSOwinSecurity

**Comment by a23o** [comment]
Dominic, it's you, it is all your fault! :) I have seen your last NDC talk on securing web APIs, and I remembered this tiny bit about access token lifetime, where you said it can be int.Max: https://vimeo.com/131635255#t=2255s 

I have long living clients, and per your suggestion I don't use refresh tokens, just a never expiring access token. (I have a client id claim in the token, so I can invalidate them server side if I want).

The problem is, we can't use int.Max for AccessTokenLifetime, because than Microsoft's bearer token validation middleware will fail the authentication. If I am correct IdentityServer3.AccessTokenValidation uses that middleware (with the UseOAuthBearerAuthentication builder method). So that's why I didn't see the client claims in the Claims Principal first place.

I did some tests and have found that with this particular middleware AccesTokenLifeTime must be smaller than (Int32.Max - UnixEpoch). I think this is funny, because this way the maximum AccesTokenLifeTime is continually decreasing. (Int32.Max - UnixEpoch) is currently around 22 years. So with this implementation the theoretical maximum of our access token lifetime is around 11 years, as of now, decreasing every second.

I never heard of the "Year 2038" problem, but apparently this is a thing: https://en.wikipedia.org/wiki/Year_2038_problem

Thank you for your help. Great talk at NDC, great software.
